In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV

In [ ]:
# Carregando o dataset
data = pd.read_csv('/content/creditcard - menor balanceado.csv')

In [ ]:
display(data.head())

Time        V1        V2        V3        V4        V5        V6  \
0   82450.0  1.314539  0.590643 -0.666593  0.716564  0.301978 -1.125467   
1   50554.0 -0.798672  1.185093  0.904547  0.694584  0.219041 -0.319295   
2   55125.0 -0.391128 -0.245540  1.122074 -1.308725 -0.639891  0.008678   
3  116572.0 -0.060302  1.065093 -0.987421 -0.029567  0.176376 -1.348539   
4   90434.0  1.848433  0.373364  0.269272  3.866438  0.088062  0.970447   

         V7        V8        V9  ...       V21       V22       V23       V24  \
0  0.388881 -0.288390 -0.132137  ... -0.170307 -0.429655 -0.141341 -0.200195   
1  0.495236  0.139269 -0.760214  ...  0.202287  0.578699 -0.092245  0.013723   
2 -0.701304 -0.027315 -2.628854  ... -0.133485  0.117403 -0.191748 -0.488642   
3  0.775644  0.134843 -0.149734  ...  0.355576  0.907570 -0.018454 -0.126269   
4 -0.721945  0.235983  0.683491  ...  0.103563  0.620954  0.197077  0.692392   

        V25       V26       V27       V28  Amount  Class  
0  0.639491  0.399476 -0.034321  0.031692    0.76      0  
1 -0.246466 -0.380057 -0.396030 -0.112901    4.18      0  
2 -0.309774  0.008100  0.163716  0.239582   15.00      0  
3 -0.339923 -0.150285 -0.023634  0.042330   57.00      0  
4 -0.206530 -0.021328 -0.019823 -0.042682    0.00      0  

[5 rows x 31 columns]

In [ ]:
display(data.isnull().sum().to_frame("Valores Nulos"))

Valores Nulos
Time                0
V1                  0
V2                  0
V3                  0
V4                  0
V5                  0
V6                  0
V7                  0
V8                  0
V9                  0
V10                 0
V11                 0
V12                 0
V13                 0
V14                 0
V15                 0
V16                 0
V17                 0
V18                 0
V19                 0
V20                 0
V21                 0
V22                 0
V23                 0
V24                 0
V25                 0
V26                 0
V27                 0
V28                 0
Amount              0
Class               0

In [ ]:
print("Dimensões do dataset:", data.shape)

Dimensões do dataset: (1759, 31)


In [ ]:
# Separando as features e a variável alvo
X = data.drop('Class', axis=1)
y = data['Class']
print("Dimensões de X (features):", X.shape)
print("Dimensões de y (alvo):", y.shape)

Dimensões de X (features): (1759, 30)
Dimensões de y (alvo): (1759,)


In [ ]:
# Dividindo em conjuntos de treino e teste
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)
print("Dimensões do conjunto de treino:", X_train.shape, y_train.shape)
print("Dimensões do conjunto de teste:", X_test.shape, y_test.shape)

Dimensões do conjunto de treino: (1231, 30) (1231,)
Dimensões do conjunto de teste: (528, 30) (528,)


In [ ]:
from sklearn.preprocessing import StandardScaler

In [ ]:
# 1. Seleção de Atributos
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)
display(pd.DataFrame(X_train_scaled, columns=X.columns).head())

Time        V1        V2        V3        V4        V5        V6  \
0  1.309896  0.240555  0.002203  0.146652 -0.670100  0.680726  0.259365   
1  0.957389  0.763281 -0.360684  0.169773 -0.312378  0.215564 -0.465452   
2 -0.617272  0.225067 -0.013784  0.594640 -0.522465  0.343885  0.067227   
3  1.540689  0.762278 -0.305273  0.063870 -0.286598  0.344209 -0.248530   
4 -0.154980  0.570407 -0.434693  0.359795 -0.456350  0.661027  2.740262   

         V7        V8        V9  ...       V20       V21       V22       V23  \
0  0.519564 -0.086447  0.339150  ... -0.359585  0.059463  0.882140 -0.193366   
1  0.367164 -0.120727  0.638538  ... -0.356637 -0.216932 -0.752393  0.389738   
2  0.440815  0.011333  0.126188  ... -0.024170 -0.203108 -0.793024 -0.094100   
3  0.341108 -0.094166  0.807486  ... -0.312942 -0.124358 -0.064109  0.086973   
4  0.055856  0.260451  0.644837  ... -0.087062  0.013946  0.468462 -0.103806   

        V24       V25       V26       V27       V28    Amount  
0 -0.613990 -0.439052 -0.438136 -0.450229  0.163161 -0.327555  
1 -0.098309 -0.463224  0.383351 -0.121446 -0.234555 -0.425766  
2 -0.634445 -0.210733  0.162614  0.251687  0.160085 -0.439243  
3 -1.267768  0.152808 -0.666009 -0.009228 -0.143259 -0.422095  
4  1.785532  0.863845 -0.552218  0.060473  0.043977 -0.377390  

[5 rows x 30 columns]

In [ ]:
from sklearn.feature_selection import SelectKBest, f_classif

In [ ]:
# Selecionando as k melhores features
k = 30
select_k_best = SelectKBest(f_classif, k=k)
X_train_kbest = select_k_best.fit_transform(X_train_scaled, y_train)
X_test_kbest = select_k_best.transform(X_test_scaled)
print("Dimensões após SelectKBest (treino):", X_train_kbest.shape)
print("Exemplo de dados de treino após SelectKBest:")
display(pd.DataFrame(X_train_kbest).head())

Dimensões após SelectKBest (treino): (1231, 30)
Exemplo de dados de treino após SelectKBest:


0         1         2         3         4         5         6   \
0  1.309896  0.240555  0.002203  0.146652 -0.670100  0.680726  0.259365   
1  0.957389  0.763281 -0.360684  0.169773 -0.312378  0.215564 -0.465452   
2 -0.617272  0.225067 -0.013784  0.594640 -0.522465  0.343885  0.067227   
3  1.540689  0.762278 -0.305273  0.063870 -0.286598  0.344209 -0.248530   
4 -0.154980  0.570407 -0.434693  0.359795 -0.456350  0.661027  2.740262   

         7         8         9   ...        20        21        22        23  \
0  0.519564 -0.086447  0.339150  ... -0.359585  0.059463  0.882140 -0.193366   
1  0.367164 -0.120727  0.638538  ... -0.356637 -0.216932 -0.752393  0.389738   
2  0.440815  0.011333  0.126188  ... -0.024170 -0.203108 -0.793024 -0.094100   
3  0.341108 -0.094166  0.807486  ... -0.312942 -0.124358 -0.064109  0.086973   
4  0.055856  0.260451  0.644837  ... -0.087062  0.013946  0.468462 -0.103806   

         24        25        26        27        28        29  
0 -0.613990 -0.439052 -0.438136 -0.450229  0.163161 -0.327555  
1 -0.098309 -0.463224  0.383351 -0.121446 -0.234555 -0.425766  
2 -0.634445 -0.210733  0.162614  0.251687  0.160085 -0.439243  
3 -1.267768  0.152808 -0.666009 -0.009228 -0.143259 -0.422095  
4  1.785532  0.863845 -0.552218  0.060473  0.043977 -0.377390  

[5 rows x 30 columns]

A utilização de métodos como SelectKBest reduziu as features para as mais informativas, diminuindo a complexidade e potencialmente melhorando a capacidade do modelo de identificar padrões associados a fraudes. Os dados de treino após essa seleção de atributos mostram que os valores foram normalizados e padronizados, facilitando o treinamento de modelos de ML e contribuindo para a precisão.

In [ ]:
from imblearn.over_sampling import SMOTE

In [ ]:
# 2. Balanceamento dos Dados
smote = SMOTE(random_state=42)
X_train_balanced, y_train_balanced = smote.fit_resample(X_train_kbest, y_train)
print("Dimensões do conjunto balanceado de treino:", X_train_balanced.shape, y_train_balanced.shape)
print("Distribuição das classes após SMOTE:")
display(pd.DataFrame(y_train_balanced, columns=['Class']).value_counts().to_frame("Quantidade"))

Dimensões do conjunto balanceado de treino: (1774, 30) (1774,)
Distribuição das classes após SMOTE:


Quantidade
Class            
0             887
1             887

O uso do SMOTE para balancear as classes ajustou o conjunto de dados para que houvesse uma distribuição igual de transações legítimas e fraudulentas (887 de cada). Essa abordagem é especialmente relevante, dado o desbalanceamento original, e foi aplicada para reduzir o viés do modelo para a classe majoritária, aumentando a sensibilidade em identificar fraudes.

In [ ]:
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier, VotingClassifier

In [ ]:
# 3. Tuning de Hiperparâmetros
param_grid_rf = {
    'n_estimators': [100, 200],
    'max_depth': [None, 10, 20],
    'min_samples_split': [2, 5]
}
param_grid_svm = {
    'C': [0.1, 1, 10],
    'kernel': ['linear', 'rbf']
}

rf = RandomForestClassifier(random_state=42)
grid_rf = GridSearchCV(rf, param_grid_rf, scoring='f1', cv=5)
grid_rf.fit(X_train_balanced, y_train_balanced)
print("Melhores parâmetros para Random Forest:", grid_rf.best_params_)

Melhores parâmetros para Random Forest: {'max_depth': None, 'min_samples_split': 5, 'n_estimators': 200}


In [ ]:
from sklearn.svm import SVC

In [ ]:
svm = SVC(probability=True, random_state=42)
grid_svm = GridSearchCV(svm, param_grid_svm, scoring='f1', cv=5)
grid_svm.fit(X_train_balanced, y_train_balanced)
print("Melhores parâmetros para SVM:", grid_svm.best_params_)

Melhores parâmetros para SVM: {'C': 10, 'kernel': 'rbf'}


In [ ]:
# Melhor modelo e parâmetros de Random Forest e SVM
best_rf = grid_rf.best_estimator_
best_svm = grid_svm.best_estimator_

Os melhores parâmetros encontrados para o Random Forest e o SVM, através de GridSearchCV ou RandomizedSearchCV, mostram a importância de ajustar os hiperparâmetros para maximizar o desempenho. A utilização de max_depth=None e min_samples_split=5 para Random Forest e C=10 para SVM, por exemplo, ajudam a modelar melhor os dados e reduzir overfitting.

In [ ]:
# 4. Ensemble de Modelos
voting_clf = VotingClassifier(estimators=[
    ('rf', best_rf),
    ('svm', best_svm),
    ('gb', GradientBoostingClassifier(random_state=42))
], voting='soft')
voting_clf.fit(X_train_balanced, y_train_balanced)

VotingClassifier(estimators=[('rf',
                              RandomForestClassifier(min_samples_split=5,
                                                     n_estimators=200,
                                                     random_state=42)),
                             ('svm',
                              SVC(C=10, probability=True, random_state=42)),
                             ('gb',
                              GradientBoostingClassifier(random_state=42))],
                 voting='soft')

A combinação de Random Forest, SVM e Gradient Boosting por meio de um Voting Classifier, utilizando a estratégia de votação soft, aumentou a robustez do modelo e permitiu o aproveitamento dos pontos fortes de cada técnica. Essa abordagem contribuiu para melhorar a precisão e a capacidade de generalização do modelo, um ponto essencial para sistemas de detecção de fraude.

In [ ]:
from sklearn.metrics import accuracy_score, f1_score, confusion_matrix, classification_report

In [ ]:
# 5. Avaliação do Modelo
y_pred = voting_clf.predict(X_test_kbest)

# Métricas de avaliação
accuracy = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred, output_dict=True)

# Exibindo os resultados como tabelas
print("\nResultados da Avaliação:")
print("Acurácia:", accuracy)
print("F1-Score (Macro):", f1)
print("Matriz de Confusão:")
display(pd.DataFrame(conf_matrix, index=["Classe 0", "Classe 1"], columns=["Previsto 0", "Previsto 1"]))
print("Relatório de Classificação:")
display(pd.DataFrame(class_report).transpose())


Resultados da Avaliação:
Acurácia: 0.9602272727272727
F1-Score (Macro): 0.9257950530035336
Matriz de Confusão:


Previsto 0  Previsto 1
Classe 0         376           4
Classe 1          17         131

Relatório de Classificação:


precision    recall  f1-score     support
0              0.956743  0.989474  0.972833  380.000000
1              0.970370  0.885135  0.925795  148.000000
accuracy       0.960227  0.960227  0.960227    0.960227
macro avg      0.963557  0.937304  0.949314  528.000000
weighted avg   0.960563  0.960227  0.959648  528.000000

As métricas de avaliação indicam um bom desempenho geral:

Acurácia: 96%, o que é alto, mas dado o cenário desbalanceado, a acurácia não é o foco principal.

F1-Score Macro: 92,6%, confirmando um bom equilíbrio entre precisão e recall para as classes.

Matriz de Confusão: O modelo prevê com precisão a maioria das transações, com 376 acertos e apenas 4 erros para a classe 0, e 131 acertos com 17 erros para a classe 1 (fraudulenta).

Relatório de Classificação: A precisão e o recall estão bem distribuídos, com um F1-score de 97,2% para a classe legítima e 92,6% para a classe fraudulenta. O recall de 88,5% na classe fraudulenta indica que o modelo está conseguindo identificar a maioria das fraudes, ainda que haja um pequeno número de falsos negativos.